src/equation_of_state/MOM_EOS.F90

```diff
@@ -134,145 +141,174 @@ logical :: Compressible = .true. !< If true, in situ density is a function of pr
-  end select
+      case (EOS_UNESCO)
+        call calculate_density_unesco(T, S, pressure, rho, start, npts, rho_ref)
+      case (EOS_WRIGHT)
+        call calculate_density_wright(T, S, pressure, rho, start, npts, rho_ref)
+      case (EOS_TEOS10)
+        call calculate_density_teos10(T, S, pressure, rho, start, npts, rho_ref)
+      case (EOS_NEMO)
+      call calculate_density_nemo(T, S, pressure, rho, start, npts, rho_ref)
+      case default
+        call MOM_error(FATAL, "calculate_density_array: EOS%form_of_EOS is not valid.")
+    end select
+  else
+    do j=start,start+npts-1 ; pres(j) = p_scale * pressure(j) ; enddo
+    select case (EOS%form_of_EOS)
+      case (EOS_LINEAR)
+        call calculate_density_linear(T, S, pres, rho, start, npts, &
+                                      EOS%Rho_T0_S0, EOS%dRho_dT, EOS%dRho_dS, rho_ref)
+      case (EOS_UNESCO)
+        call calculate_density_unesco(T, S, pres, rho, start, npts, rho_ref)
+      case (EOS_WRIGHT)
+        call calculate_density_wright(T, S, pres, rho, start, npts, rho_ref)
+      case (EOS_TEOS10)
+        call calculate_density_teos10(T, S, pres, rho, start, npts, rho_ref)
+      case (EOS_NEMO)
+      call calculate_density_nemo(T, S, pres, rho, start, npts, rho_ref)
+      case default
+        call MOM_error(FATAL, "calculate_density_array: EOS%form_of_EOS is not valid.")
+    end select
+  endif
 
-  if (present(scale)) then ; if (scale /= 1.0) then
-    do j=start,start+npts-1 ; rho(j) = scale * rho(j) ; enddo
-  endif ; endif
+  rho_scale = 1.0 ; if (present(US)) rho_scale = US%kg_m3_to_R
+  if (present(scale)) rho_scale = rho_scale * scale
+  if (rho_scale /= 1.0) then
+    do j=start,start+npts-1 ; rho(j) = rho_scale * rho(j) ; enddo
+  endif
 
 end subroutine calculate_density_array
 
-!> Calls the appropriate subroutine to calculate specific volume of sea water
-!! for scalar inputs.
-subroutine calculate_spec_vol_scalar(T, S, pressure, specvol, EOS, spv_ref, scale)
-  real,           intent(in)  :: T        !< Potential temperature referenced to the surface [degC]
-  real,           intent(in)  :: S        !< Salinity [ppt]
-  real,           intent(in)  :: pressure !< Pressure [Pa]
-  real,           intent(out) :: specvol  !< In situ? specific volume [m3 kg-1] or [R-1 ~> m3 kg-1]
-  type(EOS_type), pointer     :: EOS      !< Equation of state structure
-  real, optional, intent(in)  :: spv_ref  !< A reference specific volume [m3 kg-1].
-  real, optional, intent(in)  :: scale    !< A multiplicative factor by which to scale specific
-                                          !! volume from m3 kg-1 to the desired units [kg m-3 R-1]
+!> Calls the appropriate subroutine to calculate the density of sea water for 1-D array inputs
+!! using array extents determined from a hor_index_type.
+!! If rho_ref is present, the anomaly with respect to rho_ref is returned.
+subroutine calculate_density_HI_1d(T, S, pressure, rho, HI, EOS, US, halo)
```

Comment:
The reason for using G%HI in this way is that the existing interface is confusing and in many cases the calls to it are wrong, in that the integers being passed in would not work if we ever were to use MOM6 in global indexing mode (as we have planned to do as another test, and as SIS2 already does).  If you would feel better about it, we could also add in a 2d version as well.

Reply:
Using 2d slices instead of 1d vectors would need more memory and be slower but also more MOM6-like. I agree about the confusion w.r.t. indexing for these 1d calls but I think the new API is confusing in a different way - I was initially confused despite the documentation because HI appears as an argument. I really don't think HI should be used this way.

Reply:
A quick grep shows that there are 199 examples in the MOM6 code of the use of the horizontal indexing via the grid_type or hor_index_type for array arguments that do not have a j-dimension.  Adding the 2-d variant seems fine to me, but I do not understand what the problem is with passing 1-d arrays and using the grid or hor_index_types to indicate array extents.

Reply:
We're talking about APIs, not loops. Do you not understand that when you see `...,T,S,P,HI,...` it looks 2D in the context of the rest of MOM. How should `HI%jsd` be used or set in this context? It is setting a bad precedent.

Reply:
I have a solution to this issue that you raised (1) addresses a horrible interface to the EOS calls that we inherited from a conversion from C to Fortran almost 20 years ago, (2) avoids the direct use of HI in the calls to calculate_density, (3) greatly simplifies the interface when entire arrays are worked on, and (4) does not break existing interfaces to the equation of state for any MOM6 user code that we don't know about. I will be adding this to the PR shortly.

Reply:
The new 1d interfaces are much much nicer.